From 09b97cc503739d09496c49505b8bec82d5107549 Mon Sep 17 00:00:00 2001
From: xubing <bing.xu@intel.com>
Date: Fri, 18 Aug 2023 16:37:02 +0800
Subject: [PATCH] Fix duplicated app icons on Car Launcher

Some apps include media service, Car Launcher will draw
app icon twice if it isn't defined on
packages/apps/Car/libs/car-media-common/res/values/config.xml

Tracked-On: OAM-111463
Signed-off-by: xubing <bing.xu@intel.com>
---
 .../android/car/carlauncher/AppLauncherUtils.java    | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/src/com/android/car/carlauncher/AppLauncherUtils.java b/src/com/android/car/carlauncher/AppLauncherUtils.java
index cd8f3ec..5307b68 100644
--- a/src/com/android/car/carlauncher/AppLauncherUtils.java
+++ b/src/com/android/car/carlauncher/AppLauncherUtils.java
@@ -245,6 +245,17 @@ public class AppLauncherUtils {
                 if (shouldAddToLaunchables(componentName, appsToHide, customMediaComponents,
                         appTypes, APP_TYPE_MEDIA_SERVICES)) {
                     final boolean isDistractionOptimized = true;
+                    boolean isAvailableActivities = false;
+                    //Don't show the icon  as it will show on activities
+                    for (LauncherActivityInfo activityInfo : availableActivities) {
+                        ComponentName activitycomponentName = activityInfo.getComponentName();
+                        String activitypackageName = activitycomponentName.getPackageName();
+                        if (packageName.equals(activitypackageName)) {
+                            isAvailableActivities = true;
+                            break;
+                        }
+                    }
+                    if (isAvailableActivities) continue;
 
                     Intent intent = new Intent(Car.CAR_INTENT_ACTION_MEDIA_TEMPLATE);
                     intent.putExtra(Car.CAR_EXTRA_MEDIA_COMPONENT, componentName.flattenToString());
@@ -287,7 +298,6 @@ public class AppLauncherUtils {
                     boolean isDistractionOptimized =
                         isActivityDistractionOptimized(carPackageManager, packageName,
                             info.getName());
-
                     Intent intent = new Intent(Intent.ACTION_MAIN)
                         .setComponent(componentName)
                         .addCategory(Intent.CATEGORY_LAUNCHER)
-- 
2.34.1

